language/vm/src/abi.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    access::ModuleAccess,
+    file_format::{
+        CompiledModule, FieldDefinition, FunctionHandle, Kind, SignatureToken, StructDefinition,
+        StructFieldInformation, TypeParameterIndex,
+    },
+};
+use move_core_types::{identifier::Identifier, language_storage::ModuleId};
+
+/// Defines the Application Binary Interface (ABI) for Move modules, the `check_compatibility`
+/// function for checking ABI compatibility of two modules, and the `can_update` function for
+/// determining whether a new version of a module can safely update the old one.
+
+/// A normalized version of `SignatureToken`, a type expression appearing in struct or function
+/// declarations. Unlike `SignatureToken`s, `TypeABI`s from different modules can safely be compared.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum TypeABI {
```

Comment:
I'm not really a big fan of exposing all of these public structs/enums--especially without a clear goal. (perhaps there is a goal I'm missing)

To me, the only thing we need to expose is `pub fn can_update(old_module: &CompiledModule, new_module: &CompiledModule)`

This also makes writing tools over this a bit easier, since we can change the underlying representations without a ton of hassle

Reply:
> It should be fairly straightforward to convert an instance of ModuleABI into an LCS struct that @matbd can use to generate Rust wrappers, clients can use to decode a resource published on-chain (e.g. @runtian-zhou's resource-viewer), etc. For both of these use-cases, we might want to have a DecoratedModuleABI that augments a bytecode ModuleABI with comments, parameter names, and other info that might be useful to a client or code generator.

I think it is useful for the *ABI enums to be public in order to support this ^. E.g., a client might read a resource blob of type `M::T` and then want to extract the ABI for `M`, read the ABI for `T` from there, and then use it to deserialize the blob into an LCS representation. 

> This also makes writing tools over this a bit easier, since we can change the underlying representations without a ton of hassle

Good point. I think a good goal for these enums is to make the underlying representation of each *ABI generic enough that we never have to change it unless the bytecode format adds a backward-compatible, ABI-relevant construct (e.g., new type, public constants). Are there enums or fields here that look too brittle to support this goal? `Identifier` and `ModuleId` jump out to me as bad choices in light of that, but the others look ok to me.

Reply:
`TypeABI` and `FunctionABI` are also useful when you need to compare normalized representations of types or functions from different modules (e.g., during linking). I think this [code](https://github.com/libra/libra/blob/master/language/bytecode-verifier/src/dependencies.rs#L275) could basically be simplified to `TypeABI::new(t1) == `TypeABI::new(t2)`.

Reply:
Perhaps we should precise what "exposing" means here. Changing the (subset of) ABI interfaces that are used for code generation will force maintainers of Libra clients to (fix or more likely pull the latest version of) the codegen tools. For now, it's only us maintaining Libra clients but in the future it might be a larger community.

Reply:
>  Changing the (subset of) ABI interfaces that are used for code generation will force maintainers of Libra clients to (fix or more likely pull the latest version of) the codegen tools. 

Agreed that we should try to nail these down! The way I'm thinking of it is first defining the "raw" bytecode ABI (this PR), and then defining a "decorated ABI" that can wrap the bytecode API with metadata extracted from the source that's useful for clients (e.g., parameter names, type parameter names, comments). A Libra client should be able to codegen from a decorated ABI, and we can provide a simple utility for creating a dummy decorated ABI from a raw one.

I think of these (excellent!) YAML files you mention as defining the decorated ABI, and our task now is to figure out how to related the raw and the decorated. Some quick thoughts from skimming the YAML:

- `ArgumentABI` and `TypeArgumentABI: `name` shouldn't be part of the raw ABI (because it's not in the bytecode). But totally makes sense to include both in the decorated ABI
- The raw form of`TypeArgumentABI` should include a `KindABI`
- `ScriptABI` is a very interesting case.
  -- `name`/`doc` don't need to be part of the raw ABI (but could/should be part of the decorated ABI).
  -- In principle, the raw `ScriptABI` could be identical to the raw `FunctionABI`. The current `ScriptABI` is more restrictive in the arguments that it allows (which is reasonable, but not requred).
  -- Whether or not `code` should be part of the raw ABI is the hardest question. If we have a policy of only extending the script allowlist (i.e., never removing old script hashes), a change to `code` is not actually a breaking change (only a change to the args or type args is breaking). But I'm not sure if we want to commit ourselves to that, so treating changes to `code` as part of the raw ABI seems reasonable for no. Perhaps we could define the raw ScriptABI as raw FunctionABI + code?

Reply:
> Are there enums or fields here that look too brittle to support this goal? Identifier and ModuleId jump out to me as bad choices in light of that, but the others look ok to me.

>  I think this code could basically be simplified to `TypeABI::new(t1) == TypeABI::new(t2)`

I'm actually generally terrified of code like that. The "compatibility" question in some sense is a typing question. It looks really like classical inheritance overrides. 

```
// Expected: linking incompatible
public fun f<T: copyable>() {}
public fun f<T>() {}
```

Taking this example again, updating/"overriding" something to make it more general should be fine as it doesn't break any existing end points. And these sort of checks don't work under `==`

My worry for exposing these sorts of structs is that we lose the ability to tightly define "compatibility" as you might be doing checks with the structs/enums directly, instead of going through the normal API


--- 

I also just generally dislike APIs that look like `MyStruct::new(a, b).foo(c, d)` instead of just `my_foo(a, b, c, d)`. I think it tends to feel really leaky in terms of complexity. This distaste is likely influencing my initial reactions here :P

Reply:
I'd echo on Todd's dislike on public constructors. The way these constructors are exposed looks very similar to the old `views` we had and that resulted in a large number of clients using the file format and the views at the same time, which was fragile and took long to clean up. With that being said this may be fine... I'm not entirely sure if the two cases are comparable as views contain references and these don't.

Reply:
For visibility, some updates from the meeting:

This particular generics case is not a valid new version/override because the linker expects exact equality when comparing function signatures. That is, when a the bytecode verifier checks a function handle against the actual function definition, it expects them to match *exactly*. So there is no wiggle room here for public functions. 
---
Nonetheless, I would still have hesitations of exposing `==`, but significantly less so if we break this up into tools and passes as @dariorussi described